include/lbann/layers/learning/base_convolution.hpp

```diff
@@ -74,101 +74,87 @@ class base_convolution_layer : public learning_layer {
    *  This is this layer's contribution to the objective function
    *  gradient w.r.t. the convolutional kernel weights.
    */
-  StarMat<Dev> m_kernel_gradient;
+  StarMat<Device> m_kernel_gradient;
   /** Bias gradient.
    *  This is this layer's contribution to the objective function
    *  gradient w.r.t. the bias weights.
    */
-  StarMat<Dev> m_bias_gradient;
+  StarMat<Device> m_bias_gradient;
 
 #ifdef LBANN_HAS_CUDNN
 
   /** Convolution kernel cuDNN descriptor. */
-  cudnnFilterDescriptor_t m_kernel_cudnn_desc;
+  cudnnFilterDescriptor_t m_kernel_cudnn_desc = nullptr;
   /** Convolution cuDNN descriptor. */
-  cudnnConvolutionDescriptor_t m_convolution_cudnn_desc;
+  cudnnConvolutionDescriptor_t m_convolution_cudnn_desc = nullptr;
   /** Bias tensor cuDNN descriptor. */
-  cudnnTensorDescriptor_t m_bias_cudnn_desc;
+  cudnnTensorDescriptor_t m_bias_cudnn_desc = nullptr;
   /** Tensor cuDNN descriptors. */
   cudnn::data_parallel_layer_tensor_manager m_tensors_cudnn_desc;
 
 #endif // LBANN_HAS_CUDNN
 
 public:
 
-  base_convolution_layer(lbann_comm *comm,
-                         int num_data_dims,
-                         int num_output_channels,
-                         const std::vector<int> conv_dims,
-                         const std::vector<int> pads,
-                         const std::vector<int> strides,
-                         const std::vector<int> dilations,
+  base_convolution_layer(lbann_comm* comm,
+                         int num_data_dims, /// @todo Remove
+                         int output_channels,
+                         std::vector<int> conv_dims,
+                         std::vector<int> pads,
+                         std::vector<int> strides,
+                         std::vector<int> dilations,
                          int groups,
                          bool has_bias)
-    : learning_layer(comm),
-      m_kernel_dims(conv_dims),
-      m_kernel_size(0),
-      m_pads(pads),
-      m_strides(strides),
-      m_dilations(dilations),
-      m_num_groups(groups),
-      m_bias_scaling_factor(has_bias ? DataType(1) : DataType(0)),
-      m_kernel_gradient(this->m_comm->get_trainer_grid()),
-      m_bias_gradient(this->m_comm->get_trainer_grid())
+    : Layer(comm),
+      m_output_channels(std::max(output_channels, 1)),
```

Comment:
Why doing max with 1 here?

Reply:
It's a way to handle bad input. Although now that you mention it, it's probably better to throw an exception.

Reply:
Erroneous input should be exposed rather than hidden.